geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js

```diff
@@ -198,17 +198,141 @@ export default class ThreeDEngine {
     this.updateGroupMeshes(proxyInstances);
   }
 
+  processInstancePath (instance) {
+    if (this.meshFactory.hasVisualValue(instance)) {
+      return [instance.getInstancePath()];
+    } else if (instance.hasVisualType() || instance.getChildren().some(i => hasVisualType(i))) {
+      if (
+        instance.getType().getMetaType()
+          !== GEPPETTO.Resources.ARRAY_TYPE_NODE
+          && instance.getVisualType()
+      ) {
+        return [instance.getInstancePath()];
+      }
+      // this block keeps traversing the instances
+      if (instance.getMetaType() === GEPPETTO.Resources.INSTANCE_NODE) {
+        for (let child of instance.getChildren()) {
+          return this.processInstancePath(child);
+        }
+      } else if (
+        instance.getMetaType() === GEPPETTO.Resources.ARRAY_INSTANCE_NODE
+      ) {
+        for (let child of instance.getChildren()) {
+          return this.processInstancePath(child);
+        }
+      }
+    }
+    return [];
+  }
+
+  /*
+   * Check that the material for the already present instance did not change.
+   * return true if the color changed, otherwise false.
+   */
+
+  checkMaterial(mesh, instance) {
+    if (mesh.type === 'Mesh') {
+      if (mesh.material.color.r === instance?.color?.r
+        && mesh.material.color.g === instance?.color?.g
+        && mesh.material.color.b === instance?.color?.b
+        && mesh.material.color.opacity === instance?.color?.a) {
+          return false;
+        } else {
+          return true;
+        }
+    } else if (mesh.type === 'Group') {
+      var changed = false;
+      for (let child of mesh.children) {
+        if (this.checkMaterial(child, instance)) {
+          changed = true;
+        }
+      }
+      return changed;
+    }
+  }
+
+  updateInstanceMaterial(mesh, instance) {
+    if (mesh.type === 'Mesh') {
+      if (instance?.color?.r !== undefined
+        && instance?.color?.g !== undefined
+        && instance?.color?.b !== undefined
+        && instance?.color?.a !== undefined) {
+        for (let child of this.scene.children) {
+          if (child.instancePath === mesh.instancePath && child.uuid === mesh.uuid) {
+            child.material.color.setRGB(instance.color.r, instance.color.b, instance.color.b);
+            if (instance.color.a) {
+              child.material.transparent = true;
+              child.material.opacity = instance.color.a;
+            }
+          }
+        }
+      } else {
+        console.error("color for instance provided is malformed, please provide the color in the form");
```

Comment:
I don't think instances **have** to have a color, it can be undefined and we add a default color